### PR TITLE
Optionally install training data with libpostal

### PR DESCRIPTION
**Pull request by baduhai**
###### Description of changes

Continues the work done in PR [#179613](https://github.com/NixOS/nixpkgs/pull/179613) and fixes issue [#38024](https://github.com/NixOS/nixpkgs/issues/38024).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@Thra11 since withData is now an optional argument, closure size is only impacted when libpostalWithData is installed.

### CODE REVIEW

File: maintainers/maintainer-list.nix
Comment (cafkafk) [inline maintainers/maintainer-list.nix:1]: this should probably be its own commit, with the summary line something like `maintainer: added baduhai's matrix`

File: pkgs/development/libraries/libpostal/default.nix
Comment (cafkafk) [inline pkgs/development/libraries/libpostal/default.nix:1]: all changes in this file and `pkgs/top-level/all-packages.nix` should probably be in a single commit, with a summary line conforming to https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes (e.g. `libpostal: <brief description of changes>`)

File: maintainers/maintainer-list.nix
Comment (baduhai) [reply maintainers/maintainer-list.nix:1]: Addressed.

File: pkgs/development/libraries/libpostal/default.nix
Comment (baduhai) [reply pkgs/development/libraries/libpostal/default.nix:1]: Addressed.

**Review by Thra11 — COMMENTED** [review]
So, I attempted to test this by creating a directory with the following files:
#### main.c: (Copied from https://github.com/NixOS/nixpkgs/pull/179613#issue-1289246233)
```c
#include <stdio.h>
#include <stdlib.h>
#include <libpostal/libpostal.h>

int main(int argc, char **argv) {
    // Setup (only called once at the beginning of your program)
    if (!libpostal_setup() || !libpostal_setup_parser()) {
        exit(EXIT_FAILURE);
    }

    libpostal_address_parser_options_t options = libpostal_get_address_parser_default_options();
    libpostal_address_parser_response_t *parsed = libpostal_parse_address("781 Franklin Ave Crown Heights Brooklyn NYC NY 11216 USA", options);

    for (size_t i = 0; i < parsed->num_components; i++) {
        printf("%s: %s\n", parsed->labels[i], parsed->components[i]);
    }

    // Free parse result
    libpostal_address_parser_response_destroy(parsed);

    // Teardown (only called once at the end of your program)
    libpostal_teardown();
    libpostal_teardown_parser();
}
```
#### flake.nix:
```nix
{
  description = "libpostal-test";

  inputs.flake-utils.url = "github:numtide/flake-utils";
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/89d0c3d25e6eacc0863595852f5c8da19707697f"; # pull/240179

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      let pkgs = nixpkgs.legacyPackages.${system}; in {
        devShell = import ./shell.nix { inherit pkgs; };
      }
    );
}
```
#### shell.nix:
```nix
{ pkgs }:

let
  libpostal-with-data = pkgs.libpostal.override { withData = true; };
in
pkgs.mkShell {
  buildInputs = with pkgs; [
    libpostal-with-data
    clang
  ];
}
```
I then ran the following commands:
```
$ nix develop
$ clang main.cpp -lpostal
$ ./a.out 
ERR   Error loading transliteration module, dir=(null)
   at libpostal_setup_datadir (libpostal.c:1088) errno: No such file or directory
```
Any idea why I'm getting this error? Have you been able to build a working example?

**Comment by baduhai** [comment]
You should now be able to compile and run the example above. And the nix path-info checks out when the lib is built with data and without:

```console
 > nix path-info -Sh /nix/store/xc1755vlpwr2pa72y2565ma4x4a4nl41-libpostal-1.1
/nix/store/xc1755vlpwr2pa72y2565ma4x4a4nl41-libpostal-1.1         40.7M
 > nix path-info -Sh /nix/store/sqhmpjllh1ab0p5gzapr77mggbijs7zy-libpostal-1.1
/nix/store/sqhmpjllh1ab0p5gzapr77mggbijs7zy-libpostal-1.1          1.9G
```

The issue I'm facing now, is that even `withData = false`, nix builds `libpostal-data` even though it doesn't link it to libpostal. So the path-info doesn't increase in size, but the disk space is still used. Do you have a recommendation on how to fix that?

File: pkgs/development/libraries/libpostal/default.nix [hunk 1]
```diff
@@ -11,11 +11,43 @@ stdenv.mkDerivation rec {
...
+      mkdir -p $out/data/libpostal
+
+      ln -s ${languageClassifier}/language_classifier   $out/data/libpostal/languageClassifier
+      ln -s ${libpostalData}/transliteration            $out/data/libpostal/transliteration
+      ln -s ${libpostalData}/numex                      $out/data/libpostal/numex
+      ln -s ${libpostalData}/address_expansions         $out/data/libpostal/address_expansions
+      ln -s ${parser}/address_parser                    $out/data/libpostal/address_parser
+    '';
+  };
+
```
Comment (DeeUnderscore) [inline pkgs/development/libraries/libpostal/default.nix:45]: This should all go in a `let … in` block before `mkDerivation`. If you bind it inside of `mkDerivation`, it gets made available at build time even if not used, and so the data ends up unconditionally downloaded.

**Comment by Thra11** [comment]
Sorry it's taken me so long to get round to review this again. I can confirm that the example does now build and run successfully for me.

> The issue I'm facing now, is that even withData = false, nix builds libpostal-data even though it doesn't link it to libpostal. So the path-info doesn't increase in size, but the disk space is still used. Do you have a recommendation on how to fix that?

It should be possible to move all the optional bits outside the main mkDerivation as let bindings:
```nix
{ lib, stdenv, fetchFromGitHub, autoreconfHook, withData ? false, fetchzip }:
let
  # languageClassifier, libpostalData and parser were not release for libpostal 1.1, using data from 1.0 instead.
  languageClassifier = fetchzip {
    url = "https://github.com/openvenues/libpostal/releases/download/v1.0.0/language_classifier.tar.gz";
    sha256 = "sha256-/Gn931Nx4UDBaiFUgGqC/NJUIKQ5aZT/+OYSlcfXva8=";
    stripRoot = false;
  };

  libpostalData = fetchzip {
    url = "https://github.com/openvenues/libpostal/releases/download/v1.0.0/libpostal_data.tar.gz";
    sha256 = "sha256-FpGCkkRhVzyr08YcO0/iixxw0RK+3Of0sv/DH3GbbME=";
    stripRoot = false;
  };

  parser = fetchzip {
    url = "https://github.com/openvenues/libpostal/releases/download/v1.0.0/parser.tar.gz";
    sha256 = "sha256-OHETb3e0GtVS2b4DgklKDlrE/8gxF7XZ3FwmCTqZbqQ=";
    stripRoot = false;
  };

  data = stdenv.mkDerivation {
    name = "libpostal-data";
    buildCommand = ''
      mkdir -p $out/data/libpostal

      ln -s ${languageClassifier}/language_classifier   $out/data/libpostal/languageClassifier
      ln -s ${libpostalData}/transliteration            $out/data/libpostal/transliteration
      ln -s ${libpostalData}/numex                      $out/data/libpostal/numex
      ln -s ${libpostalData}/address_expansions         $out/data/libpostal/address_expansions
      ln -s ${parser}/address_parser                    $out/data/libpostal/address_parser
    '';
  };
in
stdenv.mkDerivation rec {
  pname = "libpostal";
  version = "1.1";
  ...

  configureFlags = [ "--disable-data-download" ]
    ++ lib.optionals stdenv.hostPlatform.isAarch64 [ "--disable-sse2" ]
    ++ lib.optionals withData [ "--datadir=${data}/data" ];
  ...
}
```
Then the derivation only refers to `data` if `withData` is true, so `data`, `languageClassifier`, `libpostalData`, and `parser` shouldn't get evaluated and fetched unless they're needed.

**Comment by baduhai** [comment]
Latest commit addresses all issues with the fix above. Hopefully it can be merged now.

**Review by Thra11 — APPROVED** [review]
Looks good to me!
Thanks @baduhai, @MasseGuillaume, and everyone else who contributed.

**Review by cafkafk — APPROVED** [review]

File: pkgs/development/libraries/libpostal/default.nix [hunk 2]
```diff
@@ -13,9 +44,9 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ autoreconfHook ];
 
-  configureFlags = [
-    "--disable-data-download"
-  ] ++ lib.optionals stdenv.hostPlatform.isAarch64 [ "--disable-sse2" ];
+  configureFlags = [ "--disable-data-download" ]
+    ++ lib.optionals stdenv.hostPlatform.isAarch64 [ "--disable-sse2" ]
+    ++ lib.optionals withData [ "--datadir=${data}/data" ];
```
Comment (doronbehar) [inline pkgs/development/libraries/libpostal/default.nix:49]: I personally liked more the previous formatting. Could you maybe settle down on this:

```suggestion
  configureFlags = [
    "--disable-data-download"
  ] ++ lib.optionals stdenv.hostPlatform.isAarch64 [
    "--disable-sse2"
  ] ++ lib.optionals withData [
    "--datadir=${data}/data"
  ];
```

This is the formatting that best handles diffs in my experience.

File: pkgs/development/libraries/libpostal/default.nix [hunk 3]
```diff
@@ -1,6 +1,37 @@
-{ lib, stdenv, fetchFromGitHub, autoreconfHook }:
+{ lib, stdenv, fetchFromGitHub, autoreconfHook, withData ? false, fetchzip }:
+let
+  # languageClassifier, libpostalData and parser were not release for libpostal 1.1, using data from 1.0 instead.
+  languageClassifier = fetchzip {
+    url = "https://github.com/openvenues/libpostal/releases/download/v1.0.0/language_classifier.tar.gz";
+    sha256 = "sha256-/Gn931Nx4UDBaiFUgGqC/NJUIKQ5aZT/+OYSlcfXva8=";
```
Comment (doronbehar) [inline pkgs/development/libraries/libpostal/default.nix:6]: Nowadays the following is more common and recommended:

```suggestion
    hash = "sha256-/Gn931Nx4UDBaiFUgGqC/NJUIKQ5aZT/+OYSlcfXva8=";
```

Same for the other hashes.

File: pkgs/development/libraries/libpostal/default.nix [hunk 4]
```diff
@@ -1,6 +1,37 @@
...
+    stripRoot = false;
```
Comment (doronbehar) [inline pkgs/development/libraries/libpostal/default.nix:7]: While at it, why is this needed?

**Comment by doronbehar** [comment]
Also, regarding your discourse comment: CI seems green and positive. Fix the merge conflict, and the small issues I mentioned and this should be good to go.
